exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -317,3 +324,259 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+		}
+	}
+
+	// If one missing but not both, return error
+	if noKind && noSelector {
+		// If proper MachinePool Machines aren't supported by the InfraMachinePool, just create Machines to match the replica count.
+		log.Info("No infraMachineKind or infraMachineSelector found, reconciling fallback machines")
+		return r.reconcileFallbackMachines(ctx, mp)
+	} else if noKind != noSelector {
+		log.Info("Only one of infraMachineKind or infraMachineSelector found, this is unexpected")
+		return errors.Errorf("only one of infraMachineKind or infraMachineSelector found, both must be present")
+	}
+
+	log.Info("Reconciling MachinePool Machines", "infrastructureMachineKind", infraMachineKind, "infrastructureMachineSelector", infraMachineSelector)
+	var infraMachineList unstructured.UnstructuredList
+
+	infraMachineList.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	infraMachineList.SetKind(infraMachineKind)
+
+	if err := r.Client.List(ctx, &infraMachineList, client.InNamespace(mp.Namespace), client.MatchingLabels(infraMachineSelector.MatchLabels)); err != nil {
+		return errors.Wrapf(err, "failed to list infra machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	if err := r.deleteDanglingMachines(ctx, mp); err != nil {
+		return errors.Wrapf(err, "failed to clean up orphan machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	if err := r.createMachinesIfNotExists(ctx, mp, infraMachineList.Items); err != nil {
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	return nil
+}
+
+// deleteDanglingMachines deletes all MachinePool Machines with an infraRef pointing to an infraMachine that no longer exists.
+// This allows us to clean up Machines when the MachinePool is scaled down.
+func (r *MachinePoolReconciler) deleteDanglingMachines(ctx context.Context, mp *expv1.MachinePool) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Deleting orphaned machines", "machinePool", mp.Name, "namespace", mp.Namespace)
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		expv1.MachinePoolNameLabel:      mp.Name,
+		clusterv1.ClusterNameLabel:      mp.Spec.ClusterName,
+		clusterv1.MachinePoolOwnedLabel: "true",
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	for i := range machineList.Items {
+		machine := &machineList.Items[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will complain if the infraRef is empty, i.e. no object kind or apiVersion.
+		if (infraRef == corev1.ObjectReference{}) {
+			log.V(2).Info("Machine has an empty infraRef, this is likely a fallback machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
```

Comment:
this seems risky, in what scenario(s) do we expect this to occur? What's the impact if we wrongly delete a machine ?

Reply:
We create fallback machines with no infraRef when we reconcile a MachinePool whose DockerMachinePool doesn't have the selector label. Until the label is set by the DockerMachinePool controller, the MachinePool controller can't tell that the DockerMachinePool supports MachinePool Machines and will make fallback machines until the label is set. When that happens, we want to be able to clean up the existing fallback machines.

Reply:
To answer the other question, there isn't any impact if we delete a fallback machine since it doesn't correspond to any resources.

Reply:
Let me ask this way: How confident are we that it's a fallback machine? The wording "this is likely a fallback machine" reads like we're not 100% confident and there could be edge cases where it's not. Deleting something is a pretty destructive action so we shouldn't do it based on guessing. If we're confident then let's reword this log to be more clear.

Reply:
I think we are confident as we expect to have fallback machines to delete. The only other case I can think of would be if the user deleted the infraRef themselves, in which case we would delete it as it falls under the case of a "dangling" machine.